.github/workflows/build_rocks.yaml

```diff
@@ -0,0 +1,107 @@
+          else
+            echo "ROCK_PATHS=$(echo "" | jq -Rsc '. / "\n" - [""]')" >> $GITHUB_ENV
+            echo "ROCKS=$(echo "" | jq -Rsc '. / "\n" - [""]')" >> $GITHUB_ENV
+          fi
+  build-rocks:
+    name: Build rock
+    runs-on: ${{ inputs.runs-on }}
+    needs: [get-rocks]
+    if: ${{ needs.get-rocks.outputs.rock-paths != '[]' }}
+    strategy:
+      matrix:
+        path: ${{ fromJSON(needs.get-rocks.outputs.rock-paths) }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Build rock
+        uses: canonical/craft-actions/rockcraft-pack@main
+        with:
+          path: ${{ matrix.path }}
+      - name: Upload ROCK to ghcr.io
+        if: ${{ !github.event.pull_request.head.repo.fork }}
+        run: |
+          IMAGE_NAME=$(yq '.name' ${{ matrix.path }}/rockcraft.yaml)
+          IMAGE_REF=${{ inputs.registry }}/${{ inputs.owner }}/$IMAGE_NAME:${{ github.run_id }}
+          skopeo --insecure-policy copy oci-archive:$(ls ${{ matrix.path }}/${{ matrix.image }}*.rock) docker://$IMAGE_REF --dest-creds "${{ github.actor }}:${{ secrets.GITHUB_TOKEN }}"
+          echo "IMAGE_NAME=$IMAGE_NAME" >> $GITHUB_ENV
+          echo "IMAGE_REF=$IMAGE_REF" >> $GITHUB_ENV
+      - name: Run Github Trivy Image Action
+        uses: aquasecurity/trivy-action@master
+        if: ${{ !github.event.pull_request.head.repo.fork }}
+        with:
+          image-ref: ${{ env.IMAGE_REF }}
+          trivy-config: ${{ inputs.trivy-image-config }}
+          exit-code: '1'
+          severity: 'CRITICAL,HIGH'
+        env:
+          TRIVY_USERNAME: ${{ github.actor }}
+          TRIVY_PASSWORD: ${{ secrets.GITHUB_TOKEN }}
+      # GitHub doesn't currently support pushing images in the Docker registry when opening a PR from a fork
+      # so an artifact is published instead
+      - name: Build rock as tarball
+        if: ${{ github.event.pull_request.head.repo.fork }}
+        run: |
+          skopeo copy oci-archive:$(ls ${{ matrix.path }}/${{ matrix.image }}*.rock) docker-archive::/tmp/$IMAGE_NAME.tar:$IMAGE_NAME:${{ github.run_id }}
+      - name: Upload image artifact
+        uses: actions/upload-artifact@v3
+        if: ${{ github.event.pull_request.head.repo.fork }}
+        with:
+          name: ${{ matrix.image }}
+          path: /tmp/${{ matrix.image }}.tar
+      # Trivy requires an OCI type tar to run the scan
+      - name: Build rock as tarball
+        if: ${{ github.event.pull_request.head.repo.fork }}
+        run: |
+          skopeo copy oci-archive:$(ls ${{ matrix.path }}/${{ matrix.image }}*.rock) dir:/tmp/$IMAGE_NAME.tar
+      - name: Run Github Trivy Image Action
+        if: ${{ github.event.pull_request.head.repo.fork }}
+        uses: aquasecurity/trivy-action@master
+        with:
+          input: ${{ matrix.image }}.tar
+          trivy-config: ${{ inputs.trivy-image-config }}
+          exit-code: '1'
+          severity: 'CRITICAL,HIGH'
```

Comment:
Isn't some factoring possible to avoid having this action twice in this workflow ?

Reply:
Perhaps. I'll try